### ISSUE TITLE
SSO Extension - Behavior/Experience Question

### ISSUE THREAD

**Issue by LibertyDev18** [AzureAD/microsoft-authentication-library-for-objc#1400]
For LOB apps configured to use the SSO extension via Authenticator, will there always be a momentary display of the in app SSO extension (white screen) modal when supplying UPN hint?  Trying to understand if there is any additional config we are missing that can avoid the momentary "blip".  These are Intune managed devices.

Question is specific to the first launch and initial token acquisition of the LOB app in our portfolio.

Thanks for your help!

**Comment by ameyapat** [comment]
Hi @LibertyDev18 , Can you provide a video recording of the momentary blip?

**Comment by LibertyDev18** [comment]
Hi @ameyapat, I've a attached a quick video snippet showing the modal that is displayed.

If we disable the SSO extension, tapping on the Authenticate button will bounce over to the Authenticator app and auth will take place within Authenticator where a similar white modal is displayed within Authenticator. Enabling the SSO extension with no code changes results in the same modal but within the LOB app.

I believe these are both Safari web auth screens and would normally show the account list but we are providing a hint so there is nothing for the user to do. In the SSO extension context is the expectation that this modal would always be briefly displayed to the user? Let me know if I can provide any additional information.

https://user-images.githubusercontent.com/16709855/140445549-a618c4ff-e081-4d95-a5e3-15cd3cfea0ac.mp4

**Comment by ameyapat** [comment]
@LibertyDev18 Thanks for the video. It looks like UI is shown when it may not be necessary. Do you see the same behavior when you don't provide login hint?
Can you go into Authenticator ->  Hamburger menu ->  Send Feedback and share the incident ID? That will give me the diagnostic data and logs for SSO extension to see what is happening and if the UI pop-up was required.

**Comment by LibertyDev18** [comment]
@ameyapat. When a hint is not supplied that screen shows the account list where a user can select the account they want to use. If they select the account, control is returned to the LOB app. BVFV2RZZ

**Comment by ameyapat** [comment]
Hi @LibertyDev18 [acquireTokenWithParameters](https://azuread.github.io/microsoft-authentication-library-for-objc/Classes/MSALPublicClientApplication.html#/Getting%20a%20token%20interactively) is meant to acquire token interactively.

When you used [acquireTokenSilentWithParameters](https://azuread.github.io/microsoft-authentication-library-for-objc/Classes/MSALPublicClientApplication.html#/c:objc(cs)MSALPublicClientApplication(im)acquireTokenSilentWithParameters:completionBlock:), it is possible for various reasons, the token service may reply with an error saying interactive login is needed. Some of the reasons could be : 
- First time signing in, token service may ask user's consent warning the use of client app.
- For security reasons, user confirmation is required for this request. Because this is an "interaction_required" error, the client should do interactive auth. This occurs because a system webview has been used to request a token for a native application - the user must be prompted to ask if this was actually the app they meant to sign into. To avoid this prompt, the redirect URI should be part of the following safe list:
http://
https://
msauth://(iOS only)
msauthv2://(iOS only)
chrome-extension:// (desktop Chrome browser only)
- Sign in frequency policies
etc...

If you want to know the exact reason, please try silent request and let me know the incident id for Authenticator after silent request.

Do you get the same behavior performing silent request after doing interactive request?

**Comment by LibertyDev18** [comment]
Hi @ameyapat, we do have msauth:// configured as the Redirect URI and yes we are able to silently acquire tokens after doing an interactive request. The interactive request resulting in just the brief display and dismiss of the white screen UI.  Stepping back just a second I'd like to confirm a couple of things around intended SSO Extension functionality.

With SSO Extension enabled for all managed apps...

- If this is the first time install of a managed LOB app in our suite of apps on the Intune managed device, will getCurrentAccountWithParameters always return nil for the MSALAccount?
- If the above is true, we must make a call to acquireTokenWithParameters which would result in the brief UI blip because acquireTokenSilentWithParameters requires an MSALAccount?
- The SSO Extension will not supply the LOB app the MSALAccount even if enrollment was done with Authenticator on the device and Authenticator has the account setup (Company Portal auth was done through Authenticator)?
- Is there a way for a managed LOB app that uses MSAL to force the bounce to Authenticator vs the in app UI blip when the SSO Extension is enabled?  When disabled the auth flow routes through Authenticator and when enabled it changes to the in app UI blip.

Purely from a LOB app perspective, unless we're missing something, it seems like the only benefit of the SSO extension is to remove the "bounce" to the Authenticator app as both paths will briefly display the white UI blip.  Depending on the answers to the questions above of course because we may be missing something.

Thanks for your help!

**Comment by ameyapat** [comment]
> If this is the first time install of a managed LOB app in our suite of apps on the Intune managed device, will getCurrentAccountWithParameters always return nil for the MSALAccount?

For first time it will return nil as we consider it not secure for apps whose trust is not evaluated to enumerate MSAL accounts. User will have to login and the initial UI blip in interactive request is required to confirm with server if the app is registered and has the right redirect URI. After 1st interactive request, It depends on if you call getCurrentAccountWithParameters synchronously or asynchronously. With synchronous MSAL will also query SSO extension for MSALAccounts, otherwise it will only query local cache. You can use account enumeration APIs : https://azuread.github.io/microsoft-authentication-library-for-objc/Classes/MSALPublicClientApplication.html#/Initializing%20MSALPublicClientApplication to query accounts to disambiguate accounts in case multiple accounts are present.

> If the above is true, we must make a call to acquireTokenWithParameters which would result in the brief UI blip because acquireTokenSilentWithParameters requires an MSALAccount?

For initial request, yes.

> The SSO Extension will not supply the LOB app the MSALAccount even if enrollment was done with Authenticator on the device and Authenticator has the account setup (Company Portal auth was done through Authenticator)?

After the first interactive request is done, you can use account enumeration APIs synchronously to get the account you want with accountIdentifier.

> Is there a way for a managed LOB app that uses MSAL to force the bounce to Authenticator vs the in app UI blip when the SSO Extension is enabled? When disabled the auth flow routes through Authenticator and when enabled it changes to the in app UI blip.

Isn't it a better experience to keep user in LOB app instead of switching UI to Authenticator? I don't think there is a way per request to use SSO extension or not. You can enable which apps can use SSO extension for all requests via https://docs.microsoft.com/en-us/azure/active-directory/develop/apple-sso-plugin#enable-sso-for-specific-apps

**Comment by LibertyDev18** [comment]
Hi @ameyapat, thank you for the additional information. Yes, agreed that the experience is better when keeping the user in the LOB app instead of switching over to Authenticator briefly.  

Just to recap, when we enable the SSO extension on a device the user experience change that we see in our LOB apps that use MSAL is the in app UI (SSO extension on) vs switch to Authenticator (SSO extension off). We have tried just about every combination (https://docs.microsoft.com/en-us/azure/active-directory/develop/apple-sso-plugin#enable-sso-for-specific-apps) of including specific apps and excluding the LOB apps and the behavior does not change if the SSO extension is enabled. So, it would appear that the user experience will be in app UI for all MSAL apps if the SSO extension is enabled?

**Comment by ameyapat** [comment]
Hi @LibertyDev18 , apologies for delayed response. Adding your app's bundle id in block list as mentioned here : https://docs.microsoft.com/en-us/azure/active-directory/develop/apple-sso-plugin#disable-sso-for-specific-apps should disable sso extension for it. Can you re-check if the correct bundle id is configured in the list? Can you provide Authenticator incident as mentioned in https://github.com/AzureAD/microsoft-authentication-library-for-objc/issues/1400#issuecomment-961575569 , the logs would show SSO extension got a request from the app but chose not to handle it as the app's bundle id was present in the block list.

**Comment by LibertyDev18** [comment]
Hi @ameyapat, it looks like we are now correctly seeing apps excluded from the SSO extension when using the block list but are seeing some interesting behavior during initial install and first attempt at an interactive request where the request is immediately rejected by the SSO extension. I just submitted feedback via Authenticator for one of these requests. Would you be able to shed some light on what the underlying error is by chance?

We believe this was previously working and may have changed with the latest, or a recent, release of the Authenticator app? 

Incident ID: XWX7XCA

Thanks for your help and any insight you can provide.

**Comment by LibertyDev18** [comment]
Hi @ameyapat, just bumping my previous question from last week to keep it on the radar.

**Comment by LibertyDev18** [comment]
Hi @ameyapat, no worries at all. I've created another incident id. The issue happens for the first interaction for an LOB app.

With the iOS SSO logging profile installed on the device I also see the following line(s) even though we are specifically making an interactive request:

**<SOAuthorizationRequest: 0x28366ca00>, authorization wants to display UI but user interaction is not enabled**

-[SOAuthorizationRequest completeWithError:] extension API called, error = Error Domain=com.apple.AuthenticationServices.AuthorizationError Code=1005 "(null)" on <SOAuthorizationRequest: 0x28366ca00>

finish authorization with error: Error Domain=com.apple.AppSSO.AuthorizationError Code=-12 "(null)"

Incident ID: P58899H3

Please let me know if I can provide any additional information. Thanks!

**Comment by ameyapat** [comment]
Hi @LibertyDev18, did your app receive an error saying interaction required : https://azuread.github.io/microsoft-authentication-library-for-objc/Enums/MSALError.html#/c:@E@MSALError@MSALErrorInteractionRequired?

**Comment by LibertyDev18** [comment]
Hi @ameyapat, no, when we make the initial call to acquireTokenWithParameters the following is returned immediately.

2022-02-25 10:42:08.070970-0600 <app_name> [604:28867] Task <4059E5C8-6293-4E25-AA86-0CA621BFB9A1>.<1> AppSSOProtocol received response (null) error Error Domain=com.apple.AuthenticationServices.AuthorizationError Code=1005 "(null)" canceling
2022-02-25 10:42:08.072029-0600 <app_name>[604:28304] Task <4059E5C8-6293-4E25-AA86-0CA621BFB9A1>.<1> finished with error [-1012] Error Domain=NSURLErrorDomain Code=-1012 "(null)" UserInfo={NSErrorFailingURLStringKey=https://login.microsoftonline.com/common/discovery/instance?api-version=1.1&authorization_endpoint=https%3A%2F%2Flogin.microsoftonline.com%2F<tenant_id>%2Foauth2%2Fv2.0%2Fauthorize, NSErrorFailingURLKey=https://login.microsoftonline.com/common/discovery/instance?api-version=1.1&authorization_endpoint=https%3A%2F%2Flogin.microsoftonline.com%2F<tenant_id>%2Foauth2%2Fv2.0%2Fauthorize, _NSURLErrorRelatedURLSessionTaskErrorKey=(
    "LocalDataTask <4059E5C8-6293-4E25-AA86-0CA621BFB9A1>.<1>"
), _NSURLErrorFailingURLSessionTaskErrorKey=LocalDataTask <4059E5C8-6293-4E25-AA86-0CA621BFB9A1>.<1>, NSUnderlyingError=0x28021d470 {Error Domain=kCFErrorDomainCFNetwork Code=-1012 "(null)" UserInfo={NSUnderlyingError=0x2802047b0 {Error Domain=com.apple.AuthenticationServices.AuthorizationError Code=1005 "(null)"}}}}

Here are the exact steps we are doing to reproduce.

1. Complete clean and reset of the iOS 15 device
2. Install Company Portal & Authenticator
3. Launch Company Portal and fully enroll the device in management (auth takes place in Authenticator)
4. Install LOB app and call to acquireTokenWithParameters fails
5. All attempts will continue to fail in the LOB app. If the user now logs into an MS app like Outlook and goes back into the LOB app, the LOB app will succeed on the call to acquireTokenWithParameters.

If we repeat the same steps above with the SSO profile removed, the acquireTokenWithParameters successfully bounces to Authenticator during step 4 on the first attempt.

**Comment by LibertyDev18** [comment]
Hi @ameyapat, hope you had a good weekend. I created a tiny single view test app to help confirm/isolate this behavior. The app has a button that calls acquireTokenWithParameters and nothing else. Following the steps in the previous post the app behaves the exact same way but there is a difference in behavior from iOS 15 to iOS 14 (14.8.1). On iOS 15, tapping the acquire token button results in an immediate failure response. On iOS 14, the auth view controller is presented in app to the user two times in succession. After the second display of the auth view, a token is successfully returned in the completion block. Same app, same code.

I have attached a video showing the iOS 14 experience. 

This behavior seems to have changed with one of the recent updates to the Authenticator app?
Why are there two presentations of the auth controller on iOS 14 in this scenario?
Are you able to confirm this behavior on your side?

https://user-images.githubusercontent.com/16709855/156047355-9361da59-6bc1-47aa-81b9-d34c395032de.mp4

**Comment by LibertyDev18** [comment]
Hi @ameyapat, sure thing. iOS 14 incident below. This had 2 UI prompts as shown above.

Incident ID: ACJGXWQ2

**Comment by LibertyDev18** [comment]
Good Afternoon @ameyapat, one additional thing to mention with the 2 UI prompts. We also see "similar" behavior with a managed 3rd party app (Zscaler) where there is an initial SSO extension prompt followed by another round trip outside of the SSO extension. This setup looks to be a bit different being that it's web view based but I don't believe the double round trip was happening there previously either. Video attached for reference in case it helps. Let me know if an incident id would be helpful for this specific flow as well.

https://user-images.githubusercontent.com/16709855/156238147-fc567b6f-743e-464c-9a1f-d8d92c9d036e.mp4

.

**Comment by ameyapat** [comment]
Hi @LibertyDev18, Thanks for the video and trying to diagnose the issue. I think we might have identified what is happening. We configure SSO extension to intercept any network requests to all our login endpoints configured in extension data set through MDM. If you app is making requests via NSURLSession, these requests by default were allowed to have the ability to present user with a webview if needed. But after iOS 15, Apple changed this behavior and by default have user interaction disabled for NSURLSession requests.
Apple is adding a property in iOS 15.4 beta to check if user interaction is allowed : https://developer.apple.com/documentation/authenticationservices/asauthorizationproviderextensionauthorizationrequest/3915982-userinterfaceenabled?changes=latest_minor&language=objc . I'll try to add a fix/workaround and will update here about its release ETA.

For the multiple prompts issue, are you making multiple calls to MSAL without getting the previous token response?

**Comment by LibertyDev18** [comment]
Hi @ameyapat, thank you for the update and additional information. Really appreciate the help with this.

For the multiple prompt issue we see the multiple prompts when making a single call to acquireTokenWithParameters. The completion block is called only once after the second prompt is displayed to the user.

We were doing some additional testing today on iOS 14 and are seeing another interesting lifecycle issue that may be related. When the SSO extension is enabled and we make a call to getCurrentAccountWithParameters it will always trigger sceneDidBecomeActive/applicationDidBecomeActive immediately before the completion block is called. This happens every time on iOS 14 with the SSO extension enabled and does not when disabled. This lifecycle event is NOT fired on iOS 15.

**Comment by ameyapat** [comment]
@LibertyDev18 After discussing with my team, the reason for the double blip is browser sso is triggered for all managed apps. To prevent this you can add your app to the block list or disable this flag : https://docs.microsoft.com/en-us/azure/active-directory/develop/apple-sso-plugin#enable-sso-for-all-managed-apps in your MDM settings.
For the -1012 issue on iOS 15, we can ship the fix in next Authenticator app release (~towards the end of this month)

**Comment by LibertyDev18** [comment]
Hi @ameyapat, thank you for the additional information and update on the Authenticator app.

With regard to the double blip, are you referring to the browser_sso_interaction_enabled flag (if we want the app to participate in SSO)? It seems that if we remove this flag then we do not get the double blip for MSAL apps on iOS 14.

Are you able to comment on the sceneDidBecomeActive (or applicationDidBecomeActive depending on OS version and implementation) lifecycle event being triggered on every call to getCurrentAccountWithParameters? It looks like this only happens on iOS 14 with the SSO extension enabled. This will lead to an infinite loop if getCurrentAccountWithParameters is called in response to the app becoming active. iOS 15 does NOT fire this lifecycle event.

1. sceneDidBecomeActive
2. getCurrentAccountWithParameters
3. sceneDidBecomeActive event fired (due to getCurrentAccountWithParameters being called at step 2) - immediately before getCurrentAccountWithParameters completion block is executed
4. getCurrentAccountWithParameters completion block executed
5. loop

**Comment by ameyapat** [comment]
@LibertyDev18 No I was referring to flag : [enable-sso-for-all-managed-apps](https://docs.microsoft.com/en-us/azure/active-directory/develop/apple-sso-plugin#enable-sso-for-all-managed-apps). If your app is not using MSAL but want SSO you can enable that flag. If your app is using MSAL you should not enable that flag. (https://docs.microsoft.com/en-us/azure/active-directory/develop/apple-sso-plugin#enable-sso-for-apps-that-dont-use-a-microsoft-identity-platform-library). Since your app is using MSAL you should disable it or add bundleId of your app in blockList. Through MSAL your app should still get SSO after that as that flag is only meant for app that don't use MSAL. Sorry for the confusion

**Comment by LibertyDev18** [comment]
Good afternoon @ameyapat. Thank you for the clarification. Okay, so we have several 3rd party apps in the portfolio that do NOT use MSAL that we DO want to enable SSO for. In this scenario we enable the enable-sso-for-all-managed-apps flag and add the bundle ids of any apps we want to exclude to the block list. The enablement/presence of the enable-sso-for-all-managed-apps flag should not adversely affect an MSAL app regardless of whether the app is in the block list correct?

Just to clarify, our testing seems to indicate that it is the combination of the **enable-sso-for-all-managed-apps** flag and the **browser_sso_interaction_enabled** flag that causes the double prompt for MSAL apps on iOS 14. When those two are set the MSAL app will double prompt unless added to the block list. **enable-sso-for-all-managed-apps** alone does not seem to create the double prompt.

Any details on the sceneDidBecomeActive events firing on calls to getCurrentAccountWithParameters? We do have logic to refresh app data on the app becoming active which can call getCurrentAccountWithParameters and be problematic (creating the infinite loop described above). We can add an additional flag that resets in the getCurrentAccountWithParameters completion block but would seem to be an errant triggering because the app does not visibly resign active when this method is called. Is MSAL and/or the SSO Extension manually triggering this event or is the system? This does not occur on iOS 15. Thanks!

**Comment by ameyapat** [comment]
Sorry for the confusion in the documentation, we will look to make it simpler. All its configuration applies to only if app is not using MSAL. See : https://docs.microsoft.com/en-us/azure/active-directory/develop/apple-sso-plugin#applications-that-dont-use-msal

If your app is using MSAL you do not need to use special configurations, see : https://docs.microsoft.com/en-us/azure/active-directory/develop/apple-sso-plugin#applications-that-use-msal.

However, as part of openid discovery, apps using MSAL initially make a network request that gets considered as request made by app not using MSAL. We will try to detect such a request and prevent prompts for MSAL apps in the future. As a workaround I'm suggesting adding MSAL app to the block list so that openid request made by MSAL app is not handled by SSO extension and cause prompt with blank screen.

browser_sso_interaction_enabled is disabled by default. It should be enabled if you want an app not using MSAL to be able to get shared credential. See : https://docs.microsoft.com/en-us/azure/active-directory/develop/apple-sso-plugin#allow-users-to-sign-in-from-unknown-applications-and-the-safari-browser. You can enable this flag if you want app not using MSAL to share credential with SSO plug-in.

Regarding sceneDidBecomeActive, in iOS 14 Apple has designed their SSO plug-in such that the plug-in for a very short amount of time comes in foreground and puts calling app in background, you might be able to observe this with the flickering bottom bar. In iOS 15 they have removed such behavior.

**Comment by LibertyDev18** [comment]
Hi @ameyapat, thank you for the additional information and all of your help with these topics.

> For the -1012 issue on iOS 15, we can ship the fix in next Authenticator app release (~towards the end of this month)

It looks like there have been two releases of the Authenticator app in the past week or so. Are you able to confirm that the fix referenced above was included in one of these releases (6.5.89 or 6.5.90)?

**Comment by ameyapat** [comment]
@LibertyDev18 The fix is included in 6.5.91. App store rollout will start on the 30th

**Comment by ameyapat** [comment]
There has been a series of hotfix deploys for Authenticator and the fix for this issue is not included in 6.5.91/6.5.92/6.5.93. I will update here when the fix for this issue is released, mostly in the 6.5.94. Apologies for the delay

**Comment by LibertyDev18** [comment]
Thank you @ameyapat. We'll pull the latest version down and a run through our test cases to confirm.

**Comment by LibertyDev18** [comment]
@ameyapat, thanks for your help with all of these topics. Testing is looking good so we should be all set for now. We can spin up another issue if/when we have additional questions.

**Comment by ameyapat** [comment]
Thanks @LibertyDev18  for reporting the issue, glad to know the fix resolved the issue.
